Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -321,18 +358,40 @@ public async Task<string> WalkTree(string treeNodeKey)
                     }
                     finally
                     {
-                        // Always call this callback, whether or not visit node was successful.
-                        await this.Parameters.Callbacks.AfterVisitNode(
+                        if (this.Parameters.CallbacksV2 != null)
+                        {
+                            // When CallbacksV2 is not null, continues with the new ITreeWalkerCallbacksV2.
+                            treeNodeContext = new TreeNodeContext(
+                                this.Parameters.SessionId,
+                                current,
+                                await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                                this.Parameters.UserContext,
+                                this.walkTreeCts.Token,
+                                this.Parameters.TreeName,
+                                this.Parameters.RootSessionId,
+                                this.currentNodeSkipActionContext
+                            );
+
+                            await this.Parameters.CallbacksV2.AfterVisitNode(treeNodeContext).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            // Always call this callback, whether or not visit node was successful.
+                            await this.Parameters.Callbacks.AfterVisitNode(
                             this.Parameters.SessionId,
                             current,
                             await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
                             this.Parameters.UserContext,
                             this.Parameters.TreeName,
                             this.Parameters.RootSessionId,
                             this.walkTreeCts.Token).ConfigureAwait(false);
+                        }
+
+                        this.currentNodeSkipActionContext = null;   // Clear the context of SkipAction and get ready for the next tree node.
```

Comment:
>// Clear the context of SkipAction and get ready for the next tree node. [](start = 68, length = 72)

Let's keep comments on their own lines.

Reply:
// Clear the context of SkipAction before visiting next node because it is only valid locally for this current tree node.

---
In reply to: [505919008](https://github.com/microsoft/Forge/pull/32#discussion_r505919008) [](ancestors = 505919008)

Reply:
replaced with your line.

---
In reply to: [505919350](https://github.com/microsoft/Forge/pull/32#discussion_r505919350) [](ancestors = 505919350,505919008)